asset-transfer-ledger-queries/chaincode-typescript/src/assetTransferLedger.ts

```diff
@@ -0,0 +1,474 @@
+        return await this.getQueryResultForQueryString(ctx, queryString);
+    }
+
+    // GetAssetsByRangeWithPagination performs a range query based on the start and end key,
+    // page size and a bookmark.
+    // The number of fetched records will be equal to or lesser than the page size.
+    // Paginated range queries are only valid for read only transactions.
+    // Example: Pagination with Range Query
+    @Transaction(false)
+    public async GetAssetsByRangeWithPagination(
+        ctx: Context,
+        startKey: string,
+        endKey: string,
+        pageSize: number,
+        bookmark: string
+    ): Promise<PaginatedQueryResult> {
+        const { iterator, metadata } =
+            await ctx.stub.getStateByRangeWithPagination(
+                startKey,
+                endKey,
+                pageSize,
+                bookmark
+            );
+        const records = await this.constructQueryResponseFromIterator(iterator);
+
+        return {
+            records,
+            fetchedRecordsCount: metadata.fetchedRecordsCount,
+            bookmark: metadata.bookmark,
+        };
+    }
+
+    // QueryAssetsWithPagination uses a query string, page size and a bookmark to perform a query
+    // for assets. Query string matching state database syntax is passed in and executed as is.
+    // The number of fetched records would be equal to or lesser than the specified page size.
+    // Supports ad hoc queries that can be defined at runtime by the client.
+    // If this is not desired, follow the QueryAssetsForOwner example for parameterized queries.
+    // Only available on state databases that support rich query (e.g. CouchDB)
+    // Paginated queries are only valid for read only transactions.
+    // Example: Pagination with Ad hoc Rich Query
+    @Transaction(false)
+    public async QueryAssetsWithPagination(
+        ctx: Context,
+        queryString: string,
+        pageSize: number,
+        bookmark: string
+    ): Promise<PaginatedQueryResult> {
+        return await this.getQueryResultForQueryStringWithPagination(
+            ctx,
+            queryString,
+            pageSize,
+            bookmark
+        );
+    }
+
+    // GetAssetHistory returns the chain of custody for an asset since issuance.
+    @Transaction(false)
+    public async GetAssetHistory(
+        ctx: Context,
+        assetID: string
+    ): Promise<HistoryQueryResult[]> {
+        const iterator = await ctx.stub.getHistoryForKey(assetID);
+        const results: HistoryQueryResult[] = [];
+        let result = await iterator.next();
+
+        while (!result.done) {
+            const historyRecord = result.value;
+            let asset: Asset | null = null;
+            if (historyRecord.value.length > 0) {
+                asset = JSON.parse(historyRecord.value.toString()) as Asset;
+            }
+            // Convert the timestamp to a Date object.
+            const seconds = historyRecord.timestamp.seconds;
```

Comment:
Timestamp includes nanos as a field, so I think it is necessary to convert this to Date as well.

```typescript
interface Timestamp {
    seconds: Long;
    nanos: number;
}
```